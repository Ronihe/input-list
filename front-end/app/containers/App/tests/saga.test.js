/**
 * Test sagas for App
 */

/* eslint-disable redux-saga/yield-effects */
//  take, call, put, select,
import { all, put } from 'redux-saga/effects';

import {
  loadInputs,
  loadInputsSuccess,
  loadInputsError,
  sendNewInput,
  sendNewInputError,
  sendNewInputSuccess,
} from '../actions';

import defaultSaga, {
  fetchInputs,
  loadInputs,
  postNewInput,
  sendNewInput,
} from '../saga';

const watcherSagas = defaultSaga();

describe('all sagas watcher', () => {
  it('should watch for other actions', () => {
    const effect = watcherSagas.next().value;
    expect(effect).toEqual(all([loadInputs(), sendNewInput()]));
  });
});

describe('fetchInputs saga', () => {
  let loadInputsGenerator;

  // have to test twice, once for a successful load and once for an error
  // the beforeEach calls all the yields that has to happen before each of the test.
  beforeEach(() => {
    loadInputsGenerator = fetchInputs();
    const callDescriptor = loadInputsGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });
  it('should dispatch the loadInputsSuccess action if it calls API successfully', () => {
    const response = [{ id: 1, message: 'hello' }, { id: 2, message: 'world' }];
    const putDescriptor = loadInputsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(loadInputsSuccess(response)));
  });
  it('should dispatch the loadMessagesError if the API call errors', () => {
    const response = new Error('some error');
    const putDescriptor = loadInputsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(loadInputsError(response)));
  });
});

describe('sendInputs saga', () => {
  let sendNewInputGenerator;

  // have to test twice, once for a successful load and once for an error
  // the beforeEach calls all the yields that has to happen before each of the test.
  beforeEach(() => {
    sendNewInputGenerator = postNewInput();
    const selectDescriptor = sendNewInputGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = sendNewInputGenerator.next('message').value;
    expect(callDescriptor).toMatchSnapshot();
  });
  it('should dispatch the sendNewMessageSuccess action if it posts to the API successfully', () => {
    const newMessage = { id: 1, message: 'hello' };
    const putDescriptor = sendNewInputGenerator.next(newMessage).value;
    expect(putDescriptor).toEqual(put(sendNewInputSuccess(newMessage)));
  });
  it('should dispatch the sendNewMessageError action if the API errors', () => {
    const response = new Error('oh no');
    const putDescriptor = sendNewInputGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(sendNewInputError(response)));
  });
});
