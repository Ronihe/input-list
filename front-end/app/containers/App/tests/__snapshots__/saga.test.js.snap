// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchInputs saga should dispatch the loadInputsSuccess action if it calls API successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3001/inputs",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`fetchInputs saga should dispatch the loadMessagesError if the API call errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3001/inputs",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`sendInputs saga should dispatch the sendNewMessageError action if the API errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`sendInputs saga should dispatch the sendNewMessageError action if the API errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3001/inputs",
      Object {
        "body": "{\\"input\\":\\"message\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`sendInputs saga should dispatch the sendNewMessageSuccess action if it posts to the API successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`sendInputs saga should dispatch the sendNewMessageSuccess action if it posts to the API successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3001/inputs",
      Object {
        "body": "{\\"input\\":\\"message\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
